src/sciml.jl

```diff
@@ -255,66 +255,168 @@ end
 """
     Matrix free operators (given by a function)
 """
-struct FunctionOperator{isinplace,T,F,Fa,Fi,P,S} <: AbstractSciMLOperator{T} # TODO
-    """ Function with signature op(u, p, t) and (optionally) op(du, u, p, t) """
+struct FunctionOperator{isinplace,T,F,Fa,Fi,Fai,Tr,P,Tt} <: AbstractSciMLOperator{T}
+    """ Function with signature op(u, p, t) and (if isinplace) op(du, u, p, t) """
     op::F
-    """ Adjoint function operator signature op(u, p, t) and (optionally) op(du, u, p, t) """
+    """ Adjoint operator"""
     op_adjoint::Fa
-    """ Adjoint function operator signature op(u, p, t) and (optionally) op(du, u, p, t) """
+    """ Inverse operator"""
     op_inverse::Fi
-    """ Size """
-    size::S
+    """ Adjoint inverse operator"""
+    op_adjoint_inverse::Fai
+    """ Traits """
+    traits::Tr
     """ Parameters """
     p::P
+    """ Time """
+    t::Tt
 
-    function FunctionOperator(op;
-                              isinplace=false,
-                              op_adjoint=nothing,
-                              op_inverse=nothing,
-                              p=nothing,
-
-                              # LinearAlgebra
-                              opnorm=nothing,
-                              isreal=true,
-                              issymmetric=false,
-                              ishermitian=false,
-                             )
-        T = eltype(op)
-
-        if LinearAlgebra.ishermitian(op) & (adjoint === nothing)
-            adjoint = op
-        end
+    function FunctionOperator(op, op_adjoint, op_inverse, op_adjoint_inverse, traits, p, t)
+        iip = traits.isinplace
+        T   = traits.T
 
-        new{isinplace,
+        new{iip,
             T,
             typeof(op),
             typeof(op_adjoint),
             typeof(op_inverse),
-            typeof(size),
+            typeof(op_adjoint_inverse),
+            typeof(traits),
             typeof(p),
+            typeof(t),
            }(
-             op, op_adjoint, op_inverse, size, p,
+             op, op_adjoint, op_inverse, op_adjoint_inverse, traits, p, t,
             )
     end
+
 end
 
-Base.size(L::FunctionOperator) = L.size
-Base.adjoint(L::FunctionOperator) = FunctionOperator(L.op_adjoint; op_inverse=L.op)
+function FunctionOperator(op, traits=nothing;
```

Comment:
It seems like this is the same as the other kwargs below?

Reply:
it's a named tuple that holds those kwargs.

```julia
# src/sciml.jl
    traits = !(traits isa Nothing) ? traits : (;
                                               opnorm = opnorm,
                                               isreal = isreal,
                                               issymmetric = issymmetric,
                                               ishermitian = ishermitian,
                                               isposdef = isposdef,


                                               isinplace = isinplace,
                                               T = T,
                                               size = size,
                                              )```

Reply:
makes it easy to pass things around eg, the definition of adjoint:

```julia
Base.size(L::FunctionOperator) = L.traits.size
function Base.adjoint(L::FunctionOperator{iip,T}) where{iip,T}

    if ishermitian(L) | (isreal(L) & issymmetric(L))
        return L
    end

    op = L.op_adjoint
    op_adjoint = L.op

    op_inverse = L.op_adjoint_inverse
    op_adjoint_inverse = L.op_inverse

    traits = (L.traits[1:end-1]..., size=reverse(size(L)))

    p = L.p
    t = L.t

    FuncitonOperator(op, op_adjoint, op_inverse, op_adjoint_inverse, traits, p, t)
end
```

Reply:
But then why have the other keyword arguments? This construction is redundant, unless it's only intended to be an internal and not used/known by users.

Reply:
the latter - need it for internals